packages/shared/src/hooks/vote/useVotePost.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  UseVotePostProps,
+  UseVotePost,
+  voteMutationHandlers,
+  UseVotePostMutationProps,
+  upvoteMutationKey,
+  VoteProps,
+  ToggleVoteProps,
+} from './types';
+
+const useVotePost = ({
+  onMutate,
+  variables,
+}: UseVotePostProps = {}): UseVotePost => {
+  const { requestMethod } = useRequestProtocol();
+  const client = useQueryClient();
+  const { user, showLogin } = useContext(AuthContext);
+  const { trackEvent } = useContext(AnalyticsContext);
+  const defaultOnMutate = ({ id, vote }) => {
+    const mutationHandler = voteMutationHandlers[vote];
+
+    if (!mutationHandler) {
+      return undefined;
+    }
+
+    const previousVote = client.getQueryData<PostData>(getPostByIdKey(id))?.post
+      ?.userState?.vote;
+    updateSinglePostCache(client, id, mutationHandler);
+
+    return () => {
+      const rollbackMutationHandler = voteMutationHandlers[previousVote];
+
+      if (!rollbackMutationHandler) {
+        return;
+      }
+
+      updateSinglePostCache(client, id, rollbackMutationHandler);
+    };
+  };
+
+  const { mutateAsync: votePost, isLoading } = useMutation(
+    ({ id, vote }: UseVotePostMutationProps) => {
+      return requestMethod(graphqlUrl, VOTE_POST_MUTATION, {
+        id,
+        vote,
+      });
+    },
+    {
+      mutationKey: variables
+        ? [...upvoteMutationKey, variables]
+        : upvoteMutationKey,
+      onMutate: onMutate ?? defaultOnMutate,
+      onError: (err, _, rollback?: () => void) => rollback?.(),
+    },
+  );
+
+  const upvotePost = useCallback(
+    ({ id }: VoteProps) => {
+      return votePost({ id, vote: UserPostVote.Up });
+    },
+    [votePost],
+  );
+  const downvotePost = useCallback(
+    ({ id }: VoteProps) => {
+      return votePost({ id, vote: UserPostVote.Down });
+    },
+    [votePost],
+  );
+  const cancelPostVote = useCallback(
+    ({ id }: VoteProps) => {
+      return votePost({ id, vote: UserPostVote.None });
+    },
+    [votePost],
+  );
+
+  const toggleUpvote = useCallback(
+    async ({ post, origin, opts }: ToggleVoteProps) => {
+      if (!post || isLoading) {
+        return;
+      }
+
+      if (!user) {
+        showLogin(AuthTriggers.Upvote);
+
+        return;
+      }
+
+      if (post?.userState?.vote === UserPostVote.Up) {
+        trackEvent(
+          postAnalyticsEvent(AnalyticsEvent.RemovePostUpvote, post, {
+            ...opts,
+            extra: { ...opts?.extra, origin },
+          }),
+        );
+
+        await cancelPostVote({ id: post.id });
+
+        return;
+      }
+
+      trackEvent(
+        postAnalyticsEvent(AnalyticsEvent.UpvotePost, post, {
+          ...opts,
+          extra: { ...opts?.extra, origin },
+        }),
+      );
+
+      await upvotePost({ id: post.id });
+    },
+    [cancelPostVote, showLogin, trackEvent, upvotePost, user, isLoading],
+  );
+
+  const toggleDownvote = useCallback(
+    async ({ post, origin, opts }: ToggleVoteProps) => {
+      if (!post || isLoading) {
```

Comment:
did we have the `isLoading` eval on the old system?
Just wondering if it can give bad UX?

Reply:
No, but it is basically the case if someone spams the button which you can do now. Don't have strong preference we can revert.

Reply:
Hmm let's see it's tricky as this doesn't have any callback to the user they might just be more frustrated lol

Reply:
To avoid not needed changes I reverted it here 8cdc429a44fdfc03a889e9e89a61646c34bfd92a